### PR TITLE
Add sign up feature

### PR DESCRIPTION
**Pull request by leandrodzp**
This PR includes:
  - sign up feature

### CODE REVIEW

File: coverage/.resultset.json [hunk 1]
```diff
@@ -1,7 +1,65 @@
 {
```
Comment (grilix) [inline coverage/.resultset.json]: I don't think we need this file in the repo. Maybe we can even remove the whole `/coverage` directory from it.

File: Gemfile [hunk 2]
```diff
@@ -17,7 +17,7 @@ gem 'figaro'
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug', platforms: [:mri, :mingw, :x64_mingw]
-  gem 'factory_girl_rails', '~> 4.5.0'
+  gem "factory_bot_rails", "~> 4.0"
```
Comment (gonzaM) [inline Gemfile]: Use single quotes

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class User < ActiveRecord::Base
+
```
Comment (gonzaM) [inline app/models/user.rb]: Remove this empty line.

File: config/application.yml [hunk 4]
```diff
@@ -1,2 +1,4 @@
 development:
-  SERVER_URL: "localhost:3000"
\ No newline at end of file
+  SERVER_URL: "localhost:3000"
+test:
+  SERVER_URL: "www.example.com"
```
Comment (grilix) [inline config/application.yml]: We should make sure every text file ends in a newline.

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class User < ActiveRecord::Base
+  
+  include DeviseTokenAuth::Concerns::User
+
+  devise :database_authenticatable, :registerable, :validatable, :trackable,
+         :recoverable, :rememberable
+
+
```
Comment (gonzaM) [inline app/models/user.rb]: Remove extra empty line. Use only one.

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+  validates :gender, :full_name, presence: true
+  
+  enum gender: [:male, :female, :other]
```
Comment (grilix) [inline app/models/user.rb]: When creating a list of keywords, we can use `%i(male female other)`.

Comment (gonzaM) [inline app/models/user.rb]: Use hash instead.
For example: `enum status: { active: 0, archived: 1 }`

File: config/application.yml (hunk 4 above)
Comment (gonzaM) [inline config/application.yml]: Add a line at the end of the file

File: spec/factories/user.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+FactoryBot.define do
+  factory :user do
+    email     { Faker::Internet.unique.email }
```
Comment (gonzaM) [inline spec/factories/user.rb]: Align:
```ruby
    ...
    password              { Faker::Internet.password(8) }
    password_confirmation { password }
    full_name             { Faker::Name.name }
    ...
```

File: spec/models/user_spec.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+require 'rails_helper'
+
+describe User do
+  describe "validations", type: :module do
```
Comment (gonzaM) [inline spec/models/user_spec.rb]: single quotes

Comment (gonzaM) [inline spec/models/user_spec.rb]: `type: model`

File: spec/requests/api/v1/users/create_spec.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+require 'rails_helper'
+
+describe 'POST api/v1/users', type: :request do
+  params = {user: FactoryBot.attributes_for(:user) }
```
Comment (grilix) [inline spec/requests/api/v1/users/create_spec.rb]: Variables shouldn't be declared here, we can use `let`:

```ruby
let(:params) { { user: FactoryBot.attributes_for(:user) } }
```

File: spec/rails_helper.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
+# This file is copied to spec/ when you run 'rails generate rspec:install'
+ENV['RAILS_ENV'] ||= 'test'
+require File.expand_path('../../config/environment', __FILE__)
+# Prevent database truncation if the environment is production
+abort('The Rails environment is running in production mode!') if Rails.env.production?
+require 'rspec/core'
+# require 'spec_helper'
```
Comment (gonzaM) [inline spec/rails_helper.rb]: You should uncomment this

File: spec/requests/api/v1/users/create_spec.rb (hunk 9 above)
Comment (gonzaM) [inline spec/requests/api/v1/users/create_spec.rb]: Let's check this test tomorrow. There are some things that you need to adjust.

File: coverage/.resultset.json (hunk 1 above)
Comment (leandrodzp) [reply coverage/.resultset.json]: I don't even know what that folder contains

**Comment by jiv** [comment]
You should add code analysis tools asap 👍

File: app/models/user.rb (hunk 6 above)
Comment (gonzaM) [reply app/models/user.rb]: No, with `%i()` the result will be: `[:male, :female, :other]`. This is an array.
You should use a hash syntax.

File: spec/rails_helper.rb (hunk 10 above)
Comment (leandrodzp) [reply spec/rails_helper.rb]: Yes I know I should but the problem is that if a do it I have a lot of dependencies errors with gems and I still don't know what that errors means. I will wait for you to came tomorrow.

File: coverage/.resultset.json (hunk 1 above)
Comment (grilix) [reply coverage/.resultset.json]: That directory stores the results of the coverage analysis. I think there's an `index.html` or something that you can open in your browser and see the stats of the coverage over the project.

These results are generated after the suite is run, and it's pretty nice for having a detailed view, but we don't need it to be in the repo.

File: spec/requests/api/v1/users/create_spec.rb [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+    it 'returns 422' do
+      post user_registration_path, params: params, as: :json
+      expect(response).to have_http_status(422)
+    end
+
+    it 'Doesn\'t create the user' do
+      expect do
+        post user_registration_path, params: params, as: :json
+        expect { User.count }.to_not change { User.count }
```
Comment (grilix) [inline spec/requests/api/v1/users/create_spec.rb]: I think the `post user_registration_path, params: params, as: :json` should be inside the `expect { .. }`.

Also, when testing for an attribute change, we can use `expect { ... }.to_not change(User, :count)` which looks cleaner IMO.

File: spec/requests/api/v1/users/create_spec.rb [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      expect do
+        post user_registration_path, params: params, as: :json
+      end.to change { User.count }.by(1)
+    end
+  end
+
+  context 'when the email is not valid' do
+    before do
+      params[:user][:email] = 'no_valid'
+    end
```
Comment (grilix) [inline spec/requests/api/v1/users/create_spec.rb]: Mutating the `let`s could lead to a harder to maintain suite. In cases like this we can, for example, just redeclare the let:

```ruby
let(:params) do
  { user: FactoryBot.attributes_for(:user, email: 'no_valid') }
end
```

**Pull request merged**